### PR TITLE
AP_Proximity: Add a low pass filter to faces

### PR DESCRIPTION
**Pull request by rishabsingh3003**
This PR adds a low pass filter to each "face". Its an atempt to reduce the reaction of glitches on Simple Avoidance. What often happens is that (as reported by several logs on the forums) there is no filtiration in the current sensor drivers, and any noise gets passed down as obstacles. This gives a sudden jerky response by the vehicle, which may even lead to crash. 
To see the filteration in action, please see this video: https://youtu.be/wmGKrU0iFSw
Here is an image from onboard logs tested on Cube Orange + Intel RealSense cameras:
image

Red line is the filtered distances, green line is the raw distances. A mode filter instead of low pass filter was also tested, but the "smothness" that comes with the low pass filter is desired. 

Additional param is also added to enable "raw distance" logging for debugging purposses. Normally it will be turned off.

### CODE REVIEW

**Comment by andyp1per** [comment]
I don't know whether it helps but there is a median low pass filkter for FFT here: https://github.com/ArduPilot/ardupilot/blob/f107ce395190f25e4189a163c2bdccefa73262b5/libraries/AP_GyroFFT/AP_GyroFFT.h#L130

**Comment by rishabsingh3003** [comment]
Thanks @andyp1per . I'd be happy to test out any other recommendations you have as well.

File: libraries/AP_Logger/LogFile.cpp [hunk 1]
```diff
@@ -607,34 +607,57 @@ void AP_Logger::Write_Proximity(AP_Proximity &proximity)
         return;
     }
 
-    AP_Proximity::Proximity_Distance_Array dist_array {};
-    proximity.get_horizontal_distances(dist_array);
+    AP_Proximity::Proximity_Distance_Array dist_array {}, filt_dist_array {};
```
Comment (peterbarker) [inline libraries/AP_Logger/LogFile.cpp:610]: We don't usually use `,` in variable declarations.

Similarly elsewhere

File: libraries/AP_Logger/LogFile.cpp [hunk 2]
```diff
@@ -607,34 +607,57 @@ void AP_Logger::Write_Proximity(AP_Proximity &proximity)
...
+        }
+        float dist_up;
+        if (!proximity.get_upward_distance(dist_up)) {
+            dist_up = 0.0f;
+        }
 
-    float dist_up;
-    if (!proximity.get_upward_distance(dist_up)) {
-        dist_up = 0.0f;
+        float close_ang = 0.0f, close_dist = 0.0f;
```
Comment (peterbarker) [inline libraries/AP_Logger/LogFile.cpp:622]:
```suggestion
        float closest_ang = 0.0f, closest_dist = 0.0f;
```

I think we can afford an extra two characters on these :-)

File: libraries/AP_Logger/LogStructure.h [hunk 3]
```diff
@@ -1370,9 +1384,10 @@ struct PACKED log_PSCZ {
 // @Field: Safety: Hardware Safety Switch status
 
 // @LoggerMessage: PRX
-// @Description: Proximity sensor data
+// @Description: Proximity Filtered sensor data
```
Comment (peterbarker) [inline libraries/AP_Logger/LogStructure.h:1387]:
```suggestion
// @Description: Filtered Proximity sensor data
```

File: libraries/AP_Logger/LogStructure.h [hunk 4]
```diff
@@ -1370,9 +1384,10 @@ struct PACKED log_PSCZ {
...
 // @Field: TimeUS: Time since system startup
-// @Field: Health: True if proximity sensor is healthy
+// @Field: Layer: Pitch(instance) at which the obstacle is at. Each layer is 30 degrees wide, 0th layer is from - 75 degrees to -45. 2nd layer captures horizontal data.
+// @Field: He: True if proximity sensor is healthy
```
Comment (peterbarker) [inline libraries/AP_Logger/LogStructure.h:1390]: This isn't actually correct.  It's incorrect in master.  But healthy comes from the `AP_Proximity_Status` structure:

```
    enum class Status {
        NotConnected = 0,
        NoData,
        Good
    };
```

So "NoData" will also be "True" :-)

File: libraries/AP_Logger/LogStructure.h [hunk 5]
```diff
@@ -1385,6 +1400,19 @@ struct PACKED log_PSCZ {
 // @Field: CAn: Angle to closest object
 // @Field: CDis: Distance to closest object
 
+// @LoggerMessage: PRXR
+// @Description: Proximity Raw sensor data
```
Comment (peterbarker) [inline libraries/AP_Logger/LogStructure.h:1404]:
```suggestion
// @Description: Raw proximity sensor data
```

File: libraries/AP_Logger/LogStructure.h [hunk 6]
```diff
@@ -1385,6 +1400,19 @@ struct PACKED log_PSCZ {
...
+// @Field: TimeUS: Time since system startup
+// @Field: Layer: Pitch(instance) at which the obstacle is at. Each layer is 30 degrees wide, 0th layer is from - 75 degrees to -45. 2nd layer captures horizontal data.
```
Comment (peterbarker) [inline libraries/AP_Logger/LogStructure.h:1406]: This is a little unclear.

I wonder if we should put the actual degrees in instead.

```suggestion
// @Field: Layer: Pitch(instance) at which the set of measurements is at. By default each layer is 30 degrees wide, 0th layer is from - 75 degrees to -45. 2nd layer captures horizontal data.
```

File: libraries/AP_Proximity/AP_Proximity.h [hunk 7]
```diff
@@ -88,6 +88,9 @@ class AP_Proximity
     // get distances in PROXIMITY_MAX_DIRECTION directions. used for sending distances to ground station
     bool get_horizontal_distances(Proximity_Distance_Array &prx_dist_array) const;
 
+    // get raw and filtered distances in 8 directions per layer. used for logging
+    bool get_active_layer_distances(uint8_t layer, AP_Proximity::Proximity_Distance_Array &prx_dist_array, AP_Proximity::Proximity_Distance_Array &prx_filt_dist_array) const;
```
Comment (peterbarker) [inline libraries/AP_Proximity/AP_Proximity.h:92]: Not sure what "active" means here....

File: libraries/AP_Proximity/AP_Proximity.h [hunk 8]
```diff
@@ -106,6 +109,9 @@ class AP_Proximity
     uint8_t get_object_count() const;
     bool get_object_angle_and_distance(uint8_t object_number, float& angle_deg, float &distance) const;
 
+    // get number of layers
+    uint8_t get_layers() const;
```
Comment (peterbarker) [inline libraries/AP_Proximity/AP_Proximity.h:113]:
```suggestion
    uint8_t get_num_layers() const;
```

File: libraries/AP_Proximity/AP_Proximity_Backend.cpp [hunk 9]
```diff
@@ -34,20 +34,14 @@ AP_Proximity_Backend::AP_Proximity_Backend(AP_Proximity &_frontend, AP_Proximity
...
-    for (uint8_t i=0; i<PROXIMITY_MAX_DIRECTION; i++) {
-        prx_dist_array.orientation[i] = i;
-        const AP_Proximity_Boundary_3D::Face face(PROXIMITY_MIDDLE_LAYER, i);
-        if (boundary.get_distance(face, prx_dist_array.distance[i])) {
-            valid_distances = true;
-        } else {
-            prx_dist_array.distance[i] = distance_max();
-        }
-    }
+    AP_Proximity::Proximity_Distance_Array prx_filt_dist_array;
```
Comment (peterbarker) [inline libraries/AP_Proximity/AP_Proximity_Backend.cpp:37]:
```suggestion
    AP_Proximity::Proximity_Distance_Array prx_filt_dist_array;  // unused
```

File: libraries/AP_Proximity/AP_Proximity_Boundary_3D.cpp [hunk 10]
```diff
@@ -280,14 +303,58 @@ uint8_t AP_Proximity_Boundary_3D::get_horizontal_object_count() const
     return PROXIMITY_NUM_SECTORS;
 }
 
+// get number of layers
+uint8_t AP_Proximity_Boundary_3D::get_layers() const
```
Comment (peterbarker) [inline libraries/AP_Proximity/AP_Proximity_Boundary_3D.cpp:307]: Not sure why this isn't in the header (and can be `constexpr` I think)

File: libraries/AP_Proximity/AP_Proximity_Boundary_3D.cpp [hunk 11]
```diff
@@ -280,14 +303,58 @@ uint8_t AP_Proximity_Boundary_3D::get_horizontal_object_count() const
...
 }
+
+// Return filtered distance for the passed in face
+bool AP_Proximity_Boundary_3D::get_filtered_distance(const Face &face, float &distance) const
+{
+    if (!face.valid()) {
+        return false;
+    }
+
+    if (_distance_valid[face.layer][face.sector]) {
```
Comment (peterbarker) [inline libraries/AP_Proximity/AP_Proximity_Boundary_3D.cpp:331]: I'd probably invert the sense of this `if` - you already have one clause returning `false`.

But it doesn't really matter.

**Review by peterbarker — CHANGES_REQUESTED** [review]
Partial review only.

File: libraries/AP_Logger/LogFile.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_Logger/LogFile.cpp:610]: let's remove the space between "dist_array" and the curly brackets.

File: libraries/AP_Logger/LogStructure.h (hunk 3 above)
Comment (rmackay9) [reply libraries/AP_Logger/LogStructure.h:1387]: I'm ok with leaving Proximity at the beginning.  I know it ends up with slightly odd english but it is good to have the subsystem at the front.

File: libraries/AP_Logger/LogStructure.h (hunk 5 above)
Comment (rmackay9) [reply libraries/AP_Logger/LogStructure.h:1404]: Again, I'd actually leave Proximity at the front..

File: libraries/AP_Proximity/AP_Proximity_Boundary_3D.cpp [hunk 12]
```diff
@@ -67,7 +71,26 @@ void AP_Proximity_Boundary_3D::add_distance(float pitch, float yaw, float distan
...
+}
+
+// Apply low pass filter on the raw distance
+void AP_Proximity_Boundary_3D::set_filtered_distance(const Face &face, float distance)
+{
+    if (!face.valid()) {
+        return;
+    }
+
+    const uint32_t update_time = AP_HAL::millis();
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_Boundary_3D.cpp:85]: often we would name this variable "now" or "now_ms".

File: libraries/AP_Proximity/AP_Proximity_Boundary_3D.h [hunk 13]
```diff
@@ -130,12 +139,20 @@ class AP_Proximity_Boundary_3D
...
+
     Vector3f _sector_edge_vector[PROXIMITY_NUM_LAYERS][PROXIMITY_NUM_SECTORS];
     Vector3f _boundary_points[PROXIMITY_NUM_LAYERS][PROXIMITY_NUM_SECTORS];
 
     float _angle[PROXIMITY_NUM_LAYERS][PROXIMITY_NUM_SECTORS];          // yaw angle in degrees to closest object within each sector and layer
     float _pitch[PROXIMITY_NUM_LAYERS][PROXIMITY_NUM_SECTORS];          // pitch angle in degrees to the closest object within each sector and layer
     float _distance[PROXIMITY_NUM_LAYERS][PROXIMITY_NUM_SECTORS];       // distance to closest object within each sector and layer
     bool _distance_valid[PROXIMITY_NUM_LAYERS][PROXIMITY_NUM_SECTORS];  // true if a valid distance received for each sector and layer
+    uint32_t _last_update_ms[PROXIMITY_NUM_LAYERS][PROXIMITY_NUM_SECTORS]; // time when distance was last updated
+    LowPassFilter<float> _filtered_distance[PROXIMITY_NUM_LAYERS][PROXIMITY_NUM_SECTORS]; // low pass filter
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_Boundary_3D.h:156]: this one can be "LowPassFilterFloat" instead of "LowPassFilter<float>"

**Comment by rmackay9** [comment]
A few small things but after those are fixed up I think we can merge this, thanks!

File: libraries/AP_Logger/LogStructure.h (hunk 3 above)
Comment (rishabsingh3003) [reply libraries/AP_Logger/LogStructure.h:1222]: Yes that was done on purpose. I remembered Randy preferred it this way..

File: libraries/AP_Proximity/AP_Proximity.h (hunk 7 above)
Comment (rishabsingh3003) [reply libraries/AP_Proximity/AP_Proximity.h:98]: It means that it has a valid distance on the layer

**Comment by rmackay9** [comment]
Merged, thanks!

**Pull request merged**